core/Decode.cpp

```diff
@@ -76,24 +130,66 @@ namespace olympia
         uint32_t num_decode = std::min(uop_queue_credits_, fetch_queue_.size());
         num_decode = std::min(num_decode, num_to_decode_);
 
+        //buffer to maximize the chances of a group match limited
+        //by max allowed latency, bounded by max group size
+        if(fusion_enable_) {
+          if(num_decode < fusion_max_group_size_
+             && latency_count < fusion_max_latency_)
+          {
+            ++latency_count;
+            return;
+          }
+        }
+
+        latency_count = 0;
+
         if(num_decode > 0)
         {
             InstGroupPtr insts =
                 sparta::allocate_sparta_shared_pointer<InstGroup>(instgroup_allocator);
+
+            InstUidListType uids;
             // Send instructions on their way to rename
             for(uint32_t i = 0; i < num_decode; ++i) {
                 const auto & inst = fetch_queue_.read(0);
                 insts->emplace_back(inst);
                 inst->setStatus(Inst::Status::DECODED);
 
+                if(fusion_enable_) uids.push_back(inst->getMavisUid());
+
                 ILOG("Decoded: " << inst);
 
                 fetch_queue_.pop();
             }
 
+            if(fusion_enable_) {
+                MatchInfoListType matches;
+                uint32_t max_itrs = 0;
+                FusionGroupContainerType & container
+                        = fuser->getFusionGroupContainer();
+                do {
+                  matchFusionGroups(matches,insts,uids,container);
+                  processMatches(matches,insts,uids);
+                  //Future feature whereIsEgon(insts,numGhosts);
+                  ++max_itrs;
+                } while(matches.size() > 0 && max_itrs < fusion_match_max_tries_);
+
+                if(max_itrs >= fusion_match_max_tries_) {
+                  throw sparta::SpartaException(
+                        "Fusion group match watch dog exceeded.");
+                }
+            }
+
+        //Debug statement
+        if(0 && fusion_enable_)  infoInsts(cout,insts);
```

Comment:
nit: tabbing funny.

Might want to throw this behind a parameter like `fusion_debug`

Reply:
tabbing fixed with clang-format.

I have added the parameter.